File: examples/cpp/estimate_nonconvex_eigenvalue.cpp
#include <iostream>
#include <proxsuite/proxqp/dense/dense.hpp> // load the dense solver backend
#include <proxsuite/proxqp/utils/random_qp_problems.hpp> // used for generating a random convex qp

using namespace proxsuite;
using namespace proxsuite::proxqp;
using T = double;

int
main()
{
  dense::isize dim = 10;
  dense::isize n_eq(dim / 4);
  dense::isize n_in(dim / 4);
  // generate a random qp
  T sparsity_factor(0.15);
  T strong_convexity_factor(1.e-2);
  dense::Model<T> qp_random = utils::dense_strongly_convex_qp(
    dim, n_eq, n_in, sparsity_factor, strong_convexity_factor);
  // make the QP nonconvex
  dense::Vec<T> diag(dim);
  diag.setOnes();
  qp_random.H.diagonal().array() -=
    2. * diag.array(); // add some nonpositive values dense matrix
  Eigen::SelfAdjointEigenSolver<dense::Mat<T>> es(qp_random.H,
                                                  Eigen::EigenvaluesOnly);
  T minimal_eigenvalue = T(es.eigenvalues().minCoeff());
  // choose scaling for regularizing default_rho accordingly
  dense::QP<T> qp(dim, n_eq, n_in); // create the QP object
  // choose the option for estimating this eigenvalue
  T estimate_minimal_eigen_value =
    dense::estimate_minimal_eigen_value_of_symmetric_matrix(
      qp_random.H, EigenValueEstimateMethodOption::ExactMethod, 1.E-6, 10000);
  bool compute_preconditioner = false;
  // input the estimate for making rho appropriate
  qp.init(qp_random.H,
          qp_random.g,
          qp_random.A,
          qp_random.b,
          qp_random.C,
          qp_random.l,
          qp_random.u,
          compute_preconditioner,
          nullopt,
          nullopt,
          nullopt,
          estimate_minimal_eigen_value);
  // print the estimates
  std::cout << "ProxQP estimate "
            << qp.results.info.minimal_H_eigenvalue_estimate << std::endl;
  std::cout << "minimal_eigenvalue " << minimal_eigenvalue << std::endl;
  std::cout << "default_rho " << qp.settings.default_rho << std::endl;
}

File: examples/python/estimate_nonconvex_eigenvalue.py
import proxsuite
import numpy as np
import scipy.sparse as spa


def generate_mixed_qp(n, seed=1, reg=-2.0):
    # A function for generating sparse random convex qps in dense format

    np.random.seed(seed)
    n_eq = int(n / 4)
    n_in = int(n / 4)
    m = n_eq + n_in

    P = spa.random(
        n, n, density=0.075, data_rvs=np.random.randn, format="csc"
    ).toarray()
    P = (P + P.T) / 2.0

    s = max(np.absolute(np.linalg.eigvals(P)))
    P += (abs(s) + 1e-02) * spa.eye(n)
    P = spa.coo_matrix(P)
    q = np.random.randn(n)
    A = spa.random(m, n, density=0.15, data_rvs=np.random.randn, format="csc").toarray()
    v = np.random.randn(n)  # Fictitious solution
    delta = np.random.rand(m)  # To get inequality
    u = A @ v
    l = -1.0e20 * np.ones(m)

    return P.toarray(), q, A[:n_eq, :], u[:n_eq], A[n_in:, :], u[n_in:], l[n_in:]


# load a qp object using qp problem dimensions
n = 10
n_eq = 2
n_in = 2
qp = proxsuite.proxqp.dense.QP(n, n_eq, n_in)
# generate a random non convex QP
H, g, A, b, C, u, l = generate_mixed_qp(n)
# initialize the model of the problem to solve
estimate_minimal_eigen_value = (
    proxsuite.proxqp.dense.estimate_minimal_eigen_value_of_symmetric_matrix(
        H, proxsuite.proxqp.EigenValueEstimateMethodOption.ExactMethod
    )
)
qp.init(H, g, A, b, C, l, u, manual_minimal_H_eigenvalue=estimate_minimal_eigen_value)
vals, _ = spa.linalg.eigs(H, which="SR")
min_eigenvalue = float(np.min(vals))
# print the estimates
print(f"{min_eigenvalue=}")
print(f"{estimate_minimal_eigen_value=}")
print(f"{qp.results.info.minimal_H_eigenvalue_estimate=}")
